src/engine/matching-engine.ts

```diff
@@ -28,93 +29,94 @@ const invalidateNonce = async (user: Hex, nonce: bigint) => {
   // TODO: Implement
 };
 
-let orderId = 0;
-
 export class MatchingEngine {
-  orders: LimitOrder[];
+  private buyOrders: Map<bigint, Map<string, LimitOrder>>;
+  private sellOrders: Map<bigint, Map<string, LimitOrder>>;
+  private orderIdToPrice: Map<string, bigint>;
 
   constructor(public readonly market: Market) {
-    this.orders = [];
+    this.buyOrders = new Map();
+    this.sellOrders = new Map();
+    this.orderIdToPrice = new Map();
   }
 
   async addOrder(order: LimitOrder) {
-    order.id = `${orderId++}`;
+    if (!order.id) {
+      order.id = randomBytes(16).toString('hex');
```

Comment:
With this, we have to explicitly check timestamps while matching, can you add that logic to the function?